mmengine/runner/runner.py

```diff
@@ -1631,26 +1629,34 @@ def train(self) -> nn.Module:
         self._train_loop = self.build_train_loop(
             self._train_loop)  # type: ignore
 
-        # `build_optimizer` should be called before `build_param_scheduler`
-        #  because the latter depends on the former
-        self.optim_wrapper = self.build_optim_wrapper(self.optim_wrapper)
-        # Automatically scaling lr by linear scaling rule
-        self.scale_lr(self.optim_wrapper, self.auto_scale_lr)
-
-        if self.param_schedulers is not None:
-            self.param_schedulers = self.build_param_scheduler(  # type: ignore
-                self.param_schedulers)  # type: ignore
-
         if self._val_loop is not None:
             self._val_loop = self.build_val_loop(
                 self._val_loop)  # type: ignore
+
         # TODO: add a contextmanager to avoid calling `before_run` many times
         self.call_hook('before_run')
 
         # initialize the model weights
         self._init_model_weights()
         # make sure checkpoint-related hooks are triggered after `before_run`
         self.load_or_resume()
+        # lazy wrap model. This should be called after `load_or_resume` to be
+        # compatible with FSDP
+        if not is_model_wrapper(self.model):
+            self.model = self.wrap_model(
+                self.cfg.get('model_wrapper_cfg'), self.model)
+
+        # `build_optimizer` should be called before `build_param_scheduler`
+        #  because the latter depends on the former
+        # `build_optimizer` should be called after `wrap_model` because
+        #  the former depends on the latter in FSDP
+        self.optim_wrapper = self.build_optim_wrapper(self.optim_wrapper)
+        # Automatically scaling lr by linear scaling rule
+        self.scale_lr(self.optim_wrapper, self.auto_scale_lr)
```

Comment:
There is a bug here when loading checkpoints. 

Currently, the `scale_lr()` method must be called before building the `ParamScheduler`. However, if there is a checkpoint, the `ParamScheduler` will be built in ``self.load_or_resume()``, before we run ``self.scale_lr(self.optim_wrapper, self.auto_scale_lr)``. 

The ``scale_lr()`` method should be modified to handle this situation.

Reply:
Thanks for your bug report & solutions! Honestly speaking, this PR does not solve all issues in FSDP integration, especially the issue with checkpoint loading. The PR stops because we find that FSDP and some other frameworks (DeepSpeed, ColossalAI, etc.) are changing the execution order of model setup, initialization, checkpoint saving/loading, which spoil the whole thing. Sometimes they may even conflict with each other. Therefore, we are working on a more elegant way to solve this issue by refactoring `Runner`, as discussed in [Discussion topic](https://github.com/open-mmlab/mmengine/discussions/749).

This refactor is working in progress, and we'll take your bug reports into consideration. Hopefully it will come out soon, and you will be able to use FSDP in MMEngine then!